packages/jupyter-ai-magics/jupyter_ai_magics/providers.py

```diff
@@ -239,6 +239,16 @@ class Config:
     provider is selected.
     """
 
+    @classmethod
```

Comment:
`@property` is only accessible on instances. `Provider` is a dual purpose construct used both as a class (representing a model provider) and as an instance (representing a model).

This two methods were added in `# class attrs` sections as they are used before we instantiate the `Provider` into a model _because_ they describe which models can be instantiated in which context.

If we used `@property` here it would lead to an exception being raised when trying to use it. Consider:

```python
class SchrödingersCat:
   @property
   def possible_states(self):
      return ['dead', 'alive']

print(SchrödingersCat.possible_states)   # <property at 0x7f79849131f0>
print(
    'The cat can be ' + ' or '.join(SchrödingersCat.possible_states)
)  # TypeError: can only join an iterable
```

What you may be suggesting (and what I initially wanted to do) is to wrap it with both `@property` and `@classmethod`, like:

 ```python
from random import choice

class SchrödingersCat:
   @classmethod
   @property
   def possible_states(self):
      return ['dead', 'alive']

   @property
   def state():
     return choice(self.possible_states)

print(SchrödingersCat.possible_states)   # ['dead', 'alive']
print(
    'The cat can be ' + ' or '.join(SchrödingersCat.possible_states)
)  # The cat can be dead or alive
```

The problem is that wrapping `@property` is no longer supported since Python 3.11 as documented in [`@classmethod`](https://docs.python.org/3.12/library/functions.html#classmethod) documentation. It still works in 3.11 and 3.12 but it will no longer work in [Python 3.13](https://docs.python.org/3.13/library/functions.html#classmethod):

> Deprecated since version 3.11, removed in version 3.13: Class methods can no longer wrap other [descriptors](https://docs.python.org/3.13/glossary.html#term-descriptor) such as [property()](https://docs.python.org/3.13/library/functions.html#property).

On the other hand, using just `@classmethod` will always work, making for much better stable API.

We could craft a custom decorator which would work as a class property descriptor, but then any downstream package would need to import such a decorator if they want to override it and we would need to maintain it. There is a number of packages which do define a custom `classproperty` method, for example django, and making one [is rather easy](https://stackoverflow.com/questions/128573/using-property-on-classmethods).

Thoughts?